The identification of novel biomarkers to diagnose and guide treatment of CRC is a priority of contemporary cancer genomics research \citep{Sveen2019}.
Biomarkers associated with immunogenic potential are of particular relevance given the emergence of immune checkpoint inhibition as an effective treatment for advanced disease.
In this study, we performed the largest unbiased assessment of transcribed but unannotated human endogenous retroviral elements in colorectal tissue to date.
We found several unique proviral loci with tumor specific over expression and many other loci with relatively high expression in healthy samples.
The tumor specific HERV sequences identified can be used to distinguish tumor from healthy and normal tumor adjacent tissue, highlighting their possible utility in screening and surveillance.
If the identified sequences are translated, they could also encode neoantigens with immunogenic potential.

Three of the seven validated HERV proviruses with tumor specific expression were of class H.
Over expression of HERV-H proviruses in CRC has been demonstrated previously with microarray \citep{Perot2012} and quantitative real time polymerase chain reaction \citep{Perot2015}.
The prominence of HERV-H transcripts in RNA-seq from CRC samples of The Cancer Genome Atlas, from which many of the samples in the current study were generated, has also been previously reported \citep{Desai2017, Attig2019}.
Therefore, our findings are generally consistent with prior work.
However, we provide new evidence of the reliability of tumor specific over expression by jointly analyzing a large number of samples from different studies.
Furthermore, our analysis was not restricted to a particular class of provirus.
The unbiased nature of our assessment allows us to contextualize the utility of measuring transcripts from three specific HERV-H loci relative to other HERV-H loci and loci from other HERV classes.

The provirus at the Xp22.32 locus appears to be the same HERV-H sequence (Xp22.3) found to be over expressed in multiple gastrointestinal cancers, including CRC, in a previous study \citep{Wentzensen2007}.
Remarkably, this sequence was subsequently found to encode an \emph{env} protein with at least three immunogenic peptides \citep{Mullins2012}.
The peptides were successfully used to sensitize human T cells that selectively attacked and killed CRC cell lines expressing the HERV Xp22.3 (Xp22.32) sequence.
Other HERV loci identified in our analysis may encode peptides with similar immunogenic potential.
The prediction of protein sequences implied by other tumor specific proviral transcripts and the testing of putative epitopes for immunogenic potential is an important goal of our future work.

This study also demonstrated the surprising frequency of HERV expression in healthy colorectal tissue.
Interpretation of this result is challenging.
HERVs have been shown to have rare functional roles in healthy tissue \citep{Rote2004} and could conceivably have a role in maintaining healthy colorectal epithelium, but this is speculative.
The over expression in healthy tissues could also be a result of the age distribution across tissue phenotypes.
Latent factor analysis is expected to adjust for multiple confounders, including subject age, but it is possible the adjustment was insufficient.
Although surprising, the unexpectedly high levels of HERV expression in healthy samples was also reassuring.
The set of reference sequences used for HERV detection in this analysis was assembled from cancer transcriptomes.
We expected to have difficulty distinguishing expression due to epigenetic de-repression of repeat sequences from an artifact.

Limitations of this study included a potential for batch effects from different studies and age confounder

These results are important because
This study is the largest to date to investigate transcriptome-wide effects of CRC on adjacent, histologically normal mucosa using RNA-seq.
The results are the first to demonstrate the consistent overexpression of 20 tumor-associated genes in histologically normal tissue sampled adjacent to tumors.
Remarkably, genes contributing to established oncogenic pathways, such as epidermal growth fact
