In this study, we performed the largest unbiased assessment of transcribed but unannotated human endogenous retroviral elements in colorectal tissue to date.
We found several unique proviral loci with tumor specific over expression and many other loci with relatively high expression in healthy samples.
Human endogenous retroviral elements are normally transcriptionally repressed repetitive sequences that are often de-repressed in epigenetically altered cancer cells \citep{Babaian2016}.
HERV transcripts have been shown to engage pattern recognition receptors and activate interferon responses in cancer cell lines \citep{Chiappinelli2015, Roulois2015} and have been associated with tumor infiltration by T regulatory cells \citep{Desai2017}, response to PD-L1 blockade in urothelial carcinoma, and survival in subjects with CRC \citep{Solovyov2018}.
Our results are important because the tumor specific HERV sequences we identified could be used to distinguish tumor from non-tumor tissue for screening and surveillance and to predict or even modify response to immunotherapies.

Three of the seven HERV proviruses we validated as tumor specific were from group H.
Over expression of HERV-H proviruses in CRC has been demonstrated previously with microarray \citep{Perot2012} and quantitative real time polymerase chain reaction \citep{Perot2015}.
The prominence of HERV-H transcripts in RNA-seq from CRC samples of The Cancer Genome Atlas, from which many of the samples in the current study were generated, has also been previously reported \citep{Desai2017, Attig2019}.
Therefore, our findings are generally consistent with prior work.
However, we provide new evidence of the reliability of tumor specific over expression by jointly analyzing a large number of samples from different studies.
Furthermore, our analysis was not restricted to a particular group of proviruses.
The unbiased nature of our assessment provides important context for the community of researchers selecting genomic features to develop as biomarkers and immunotargets for CRC.

We note that the tumor specific provirus we mapped to the high resolution cytoband Xp22.32 appears to be the same HERV-H provirus previously mapped to the lower resolution cytoband Xp22.3 and found to be over expressed in CRC \citep{Wentzensen2007}.
Remarkably, this sequence was subsequently found to encode an \emph{env} protein including at least three immunogenic epitopes \citep{Mullins2012}.
The epitopes were successfully used to sensitize human cytotoxic T cells in culture, and the sensitized lymphocytes were shown to selectively lyse CRC cell lines expressing the HERV Xp22.3 provirus.
Other HERV loci identified in our analysis may encode peptides with similar immunogenic potential.
Identification of protein sequences that could be translated from tumor specific proviral transcripts is therefore an important goal of our future work.

This study also demonstrated the surprising frequency of HERV expression in healthy colorectal tissue.
Interpretation of this result is challenging.
HERVs have been shown to have rare functional roles in healthy cells \citep{Rote2004} and could conceivably have a role in maintaining healthy colorectal epithelium, but this is speculative.
Over expression in our healthy samples could also result from technical (i.e. sequencing protocols) or demographic (i.e. age) differences between phenotypes.
Latent factor analysis is expected to adjust for multiple confounders, including subject age and sequencing artifacts, but the adjustment may be insufficient.
Although surprising, the detection of high levels of HERV expression in healthy samples reduced the likelihood that using a reference HERV transcriptome assembled exclusively from contigs of cancer cells had biased our results.

The major limitation of this study was its reliance on complex bioinformatic and statistical methods to aggregate heterogeneous RNA-seq datasets and quantify poorly annotated transcripts.
Specifically, the definition and quantification of HERV elements was of critical importance.
We used a published, hidden Markov model based annotation of a \emph{de novo} assembled cancer transcriptome to define and locate the HERV transcripts and loci we studied \citep{Attig2019}.
We used a published, alignment-free, k-mer matching quantification algorithm to probabilistically estimate transcript abundance of repetitive sequences that may otherwise have been excluded from analysis due to multiple possible mappings \citep{Patro2017}.
We used state of the art statistical packages designed specifically for high dimensional genomics analyses to perform association tests \citep{Love2014, sva}.
We also pursued multiple validations of our findings and methods to ensure reproducibility.

The aim of this work was to prioritize HERV transcripts for potential development as novel biomarkers or immunotargets in the management of CRC.
Markers and targets for immune modulation are of particular relevance given the emergence of immune checkpoint inhibition as an effective treatment for advanced CRC.
The seven proviruses we identified were consistently and reliably over expression in colorectal tumors and warrant future investigation.
